File: backend/src/services/syncQueue.service.js
/**
 * Phase 3.4.0: Sync Queue Service
 * Phase 3.5.2: Enhanced with better conflict resolution and queue optimization
 * Manages sync queue with priority, retry, and conflict resolution
 */

import mongoose from 'mongoose';
import SyncQueue from '../models/SyncQueue.js';
import { syncOfflineData } from './sync.service.js';
import { detectConflict, resolveConflict as resolveConflictWithStrategy, RESOLUTION_STRATEGIES, getDefaultStrategy } from './conflictResolution.service.js';
import QuizResult from '../models/QuizResult.js';
import DrillLog from '../models/DrillLog.js';
import GameScore from '../models/GameScore.js';
import logger from '../config/logger.js';

/**
 * Add item to sync queue
 * Phase 3.5.2: Enhanced with duplicate detection
 * @param {string} userId - User ID
 * @param {string} dataType - Type of data ('quiz', 'drill', 'game', 'module')
 * @param {Object} payload - Data to sync
 * @param {Object} options - Options (priority, metadata)
 */
export const addToSyncQueue = async (userId, dataType, payload, options = {}) => {
  try {
    // Phase 3.5.2: Check for duplicate pending items to avoid duplicates
    const existingItem = await SyncQueue.findOne({
      userId,
      dataType,
      status: { $in: ['pending', 'processing'] },
      // Check if same data (basic duplicate detection)
      'payload.userId': payload.userId || userId,
      'payload.moduleId': payload.moduleId,
      'payload.drillId': payload.drillId,
      'payload.completedAt': payload.completedAt,
    });

    if (existingItem) {
      logger.debug(`Duplicate sync item detected, skipping: ${dataType} for user ${userId}`);
      return existingItem; // Return existing item
    }

    const queueItem = await SyncQueue.create({
      userId,
      institutionId: options.institutionId || null,
      dataType,
      payload,
      priority: options.priority || getDefaultPriority(dataType),
      metadata: {
        deviceId: options.deviceId,
        appVersion: options.appVersion,
        syncVersion: options.syncVersion || '1.0',
        createdAt: options.createdAt || new Date(),
      },
    });

    logger.info(`Added ${dataType} to sync queue for user ${userId} (priority: ${queueItem.priority})`);
    return queueItem;
  } catch (error) {
    logger.error('Add to sync queue error:', error);
    throw error;
  }
};

/**
 * Get default priority for data type
 * Phase 3.5.2: Priority-based queue management
 */
const getDefaultPriority = (dataType) => {
  const priorities = {
    'drill': 1,      // Highest priority - drills are time-sensitive
    'quiz': 3,       // Medium-high priority
    'game': 5,       // Medium priority
    'module': 7,     // Lower priority - modules are downloaded, not synced
  };
  return priorities[dataType] || 5;
};

/**
 * Process sync queue (process N items in priority order)
 * Phase 3.5.2: Enhanced with better conflict resolution and performance
 * @param {string} userId - User ID (optional, if not provided processes all users)
 * @param {number} batchSize - Number of items to process
 * @param {Object} options - Processing options
 * @returns {Object} Processing results
 */
export const processSyncQueue = async (userId = null, batchSize = 10, options = {}) => {
  try {
    const { autoResolveConflicts = true, conflictStrategy = null } = options;

    // Find pending items, ordered by priority (lower number = higher priority) and creation time
    const query = { status: 'pending' };
    if (userId) {
      query.userId = userId;
    }

    // Phase 3.5.2: Optimize query with better indexing and limits
    const pendingItems = await SyncQueue.find(query)
      .sort({ priority: 1, createdAt: 1 })
      .limit(batchSize)
      .lean();

    if (pendingItems.length === 0) {
      return {
        processed: 0,
        synced: 0,
        failed: 0,
        conflicts: 0,
      };
    }

    const results = {
      processed: 0,
      synced: 0,
      failed: 0,
      conflicts: 0,
      autoResolved: 0,
    };

    // Phase 3.5.2: Group items by data type for batch processing (more efficient)
    const groupedByType = {};
    for (const item of pendingItems) {
      if (!groupedByType[item.dataType]) {
        groupedByType[item.dataType] = [];
      }
      groupedByType[item.dataType].push(item);
    }

    // Process each data type in batches
    for (const [dataType, items] of Object.entries(groupedByType)) {
      try {
        // Mark items as processing
        const itemIds = items.map((item) => item._id);
        await SyncQueue.updateMany(
          { _id: { $in: itemIds } },
          {
            status: 'processing',
            lastAttemptAt: new Date(),
            $inc: { retryCount: 1 },
          }
        );

        // Phase 3.5.2: Check for conflicts before syncing
        const itemsToSync = [];
        const conflictItems = [];

        for (const item of items) {
          // Try to detect conflicts by checking existing data
          try {
            const conflict = await checkForConflicts(item, dataType);
            if (conflict.hasConflict && autoResolveConflicts) {
              // Auto-resolve if possible
              const strategy = conflictStrategy || conflict.recommendedStrategy || getDefaultStrategy(dataType);
              const resolution = resolveConflictWithStrategy(conflict, strategy);

              if (resolution.resolved) {
                // Update payload with resolved data
                item.payload = resolution.data;
                itemsToSync.push(item);
                results.autoResolved++;
                logger.debug(`Auto-resolved conflict for ${dataType} item ${item._id}`);
              } else {
                conflictItems.push({ item, conflict, resolution });
              }
            } else if (conflict.hasConflict) {
              conflictItems.push({ item, conflict });
            } else {
              itemsToSync.push(item);
            }
          } catch (error) {
            // If conflict detection fails, try to sync anyway
            logger.warn(`Conflict detection failed for item ${item._id}:`, error.message);
            itemsToSync.push(item);
          }
        }

        // Handle conflicts that need manual resolution
        for (const { item, conflict, resolution } of conflictItems) {
          await SyncQueue.updateOne(
            { _id: item._id },
            {
              status: 'conflict',
              error: conflict.message || 'Conflict detected',
              'conflictData.serverData': conflict.serverData,
              'conflictData.localData': conflict.clientData,
              'conflictData.resolution': resolution?.requiresManualResolution ? 'manual' : (conflict.recommendedStrategy || 'manual'),
            }
          );
          results.conflicts++;
        }

        // Prepare sync payload for items without conflicts
        const syncData = {
          quizzes: [],
          drillLogs: [],
          games: [],
          modules: [],
        };

        for (const item of itemsToSync) {
          if (item.dataType === 'quiz') {
            syncData.quizzes.push(item.payload);
          } else if (item.dataType === 'drill') {
            syncData.drillLogs.push(item.payload);
          } else if (item.dataType === 'game') {
            syncData.games.push(item.payload);
          } else if (item.dataType === 'module') {
            syncData.modules.push(item.payload);
          }
        }

        // Sync the data
        if (itemsToSync.length > 0) {
          const userIdForSync = itemsToSync[0].userId;
          const syncResult = await syncOfflineData(userIdForSync, syncData);

          // Update queue items based on results
          let quizIndex = 0;
          let drillIndex = 0;
          let gameIndex = 0;

          for (const item of itemsToSync) {
            results.processed++;
            let itemSynced = false;

            if (item.dataType === 'quiz' && syncResult.quizzes.synced > quizIndex) {
              itemSynced = true;
              quizIndex++;
            } else if (item.dataType === 'drill' && syncResult.drillLogs.synced > drillIndex) {
              itemSynced = true;
              drillIndex++;
            } else if (item.dataType === 'game' && syncResult.games.synced > gameIndex) {
              itemSynced = true;
              gameIndex++;
            }

            if (itemSynced) {
              await SyncQueue.updateOne(
                { _id: item._id },
                {
                  status: 'synced',
                  syncedAt: new Date(),
                }
              );
              results.synced++;
            } else {
              // Check if retry limit reached
              const updatedItem = await SyncQueue.findById(item._id);
              if (updatedItem.retryCount >= updatedItem.maxRetries) {
                await SyncQueue.updateOne(
                  { _id: item._id },
                  {
                    status: 'failed',
                    error: 'Max retries exceeded',
                  }
                );
                results.failed++;
              } else {
                // Reset to pending for retry with exponential backoff
                const backoffDelay = Math.min(300000, Math.pow(2, updatedItem.retryCount) * 60000); // Max 5 minutes
                await SyncQueue.updateOne(
                  { _id: item._id },
                  {
                    status: 'pending',
                    $set: {
                      'metadata.retryAfter': new Date(Date.now() + backoffDelay),
                    },
                  }
                );
              }
            }
          }

          // Handle conflicts from sync result
          if (syncResult.conflicts && syncResult.conflicts.length > 0) {
            for (const conflict of syncResult.conflicts) {
              results.conflicts++;
              // Items already handled above, just log
              logger.warn(`Conflict during sync: ${conflict.type} - ${conflict.message}`);
            }
          }
        }
      } catch (error) {
        logger.error(`Error processing ${dataType} items:`, error);
        // Mark items as failed or reset for retry
        for (const item of items) {
          const updatedItem = await SyncQueue.findById(item._id);
          if (updatedItem.retryCount >= updatedItem.maxRetries) {
            await SyncQueue.updateOne(
              { _id: item._id },
              {
                status: 'failed',
                error: error.message,
              }
            );
            results.failed++;
          } else {
            // Exponential backoff
            const backoffDelay = Math.min(300000, Math.pow(2, updatedItem.retryCount) * 60000);
            await SyncQueue.updateOne(
              { _id: item._id },
              {
                status: 'pending',
                error: error.message,
                $set: {
                  'metadata.retryAfter': new Date(Date.now() + backoffDelay),
                },
              }
            );
          }
        }
      }
    }

    logger.info(
      `Processed sync queue: ${results.processed} items, ${results.synced} synced, ${results.failed} failed, ${results.conflicts} conflicts, ${results.autoResolved} auto-resolved`
    );

    return results;
  } catch (error) {
    logger.error('Process sync queue error:', error);
    throw error;
  }
};

/**
 * Check for conflicts before syncing
 * Phase 3.5.2: Enhanced conflict detection
 */
const checkForConflicts = async (queueItem, dataType) => {
  try {
    const { userId, payload } = queueItem;

    // Try to find existing data that might conflict
    let existingData = null;

    switch (dataType) {
      case 'quiz':
        if (payload.moduleId && payload.completedAt) {
          existingData = await QuizResult.findOne({
            userId,
            moduleId: payload.moduleId,
            completedAt: {
              $gte: new Date(new Date(payload.completedAt).getTime() - 60000), // Within 1 minute
              $lte: new Date(new Date(payload.completedAt).getTime() + 60000),
            },
          }).lean();
        }
        break;

      case 'drill':
        if (payload.drillId && payload.participatedAt) {
          existingData = await DrillLog.findOne({
            userId,
            drillId: payload.drillId,
            participatedAt: {
              $gte: new Date(new Date(payload.participatedAt).getTime() - 60000),
              $lte: new Date(new Date(payload.participatedAt).getTime() + 60000),
            },
          }).lean();
        }
        break;

      case 'game':
        if (payload.gameType && payload.completedAt) {
          existingData = await GameScore.findOne({
            userId,
            gameType: payload.gameType,
            completedAt: {
              $gte: new Date(new Date(payload.completedAt).getTime() - 60000),
              $lte: new Date(new Date(payload.completedAt).getTime() + 60000),
            },
          }).lean();
        }
        break;
    }

    if (existingData) {
      // Convert MongoDB document to plain object if needed
      const serverData = existingData._id ? { ...existingData, _id: existingData._id.toString() } : existingData;
      const clientData = payload;

      return detectConflict(serverData, clientData, dataType);
    }

    return { hasConflict: false };
  } catch (error) {
    logger.warn(`Conflict check error for ${dataType}:`, error.message);
    return { hasConflict: false }; // If check fails, assume no conflict
  }
};

/**
 * Resolve conflict manually
 * Phase 3.5.2: Enhanced with new conflict resolution strategies
 * @param {string} queueItemId - Queue item ID
 * @param {string} resolution - Resolution strategy
 * @param {Object} resolvedData - Resolved data (if merge or client-wins)
 */
export const resolveConflict = async (queueItemId, resolution, resolvedData = null) => {
  try {
    const queueItem = await SyncQueue.findById(queueItemId);
    if (!queueItem) {
      throw new Error('Queue item not found');
    }

    if (queueItem.status !== 'conflict') {
      throw new Error('Item is not in conflict state');
    }

    // Use enhanced conflict resolution service
    const conflict = {
      serverData: queueItem.conflictData?.serverData,
      clientData: queueItem.payload,
      type: queueItem.conflictData?.type || 'generic',
    };

    const resolutionResult = resolveConflictWithStrategy(conflict, resolution, {
      mergedData: resolvedData,
    });

    if (!resolutionResult.resolved && resolutionResult.requiresManualResolution) {
      // Still needs manual resolution
      return {
        success: false,
        message: 'Item requires manual resolution',
        conflict: conflict,
      };
    }

    let finalData = resolutionResult.data;

    if (resolution === RESOLUTION_STRATEGIES.SERVER_WINS) {
      // Use server data (already in database, don't sync)
      await SyncQueue.updateOne(
        { _id: queueItemId },
        {
          status: 'synced',
          syncedAt: new Date(),
          'conflictData.resolution': resolution,
        }
      );
      return { success: true, message: 'Conflict resolved: server data used' };
    }

    // Try to sync again with resolved data
    try {
      const syncData = {};
      if (queueItem.dataType === 'quiz') {
        syncData.quizzes = [finalData];
      } else if (queueItem.dataType === 'drill') {
        syncData.drillLogs = [finalData];
      } else if (queueItem.dataType === 'game') {
        syncData.games = [finalData];
      }

      // Use updateOne instead of insertMany for conflicts
      await syncOfflineData(queueItem.userId, syncData);

      await SyncQueue.updateOne(
        { _id: queueItemId },
        {
          status: 'synced',
          syncedAt: new Date(),
          'conflictData.resolution': resolution,
        }
      );

      return { success: true, message: 'Conflict resolved and synced successfully' };
    } catch (syncError) {
      await SyncQueue.updateOne(
        { _id: queueItemId },
        {
          status: 'failed',
          error: syncError.message,
          'conflictData.resolution': resolution,
        }
      );
      throw syncError;
    }
  } catch (error) {
    logger.error('Resolve conflict error:', error);
    throw error;
  }
};

/**
 * Get queue status for a user
 * Phase 3.5.2: Enhanced with more statistics
 * @param {string} userId - User ID
 * @returns {Object} Queue status
 */
export const getQueueStatus = async (userId) => {
  try {
    const userIdObj = typeof userId === 'string' ? new mongoose.Types.ObjectId(userId) : userId;
    
    const counts = await SyncQueue.aggregate([
      { $match: { userId: userIdObj } },
      {
        $group: {
          _id: '$status',
          count: { $sum: 1 },
        },
      },
    ]);

    const statusCounts = {
      pending: 0,
      processing: 0,
      synced: 0,
      failed: 0,
      conflict: 0,
    };

    for (const item of counts) {
      statusCounts[item._id] = item.count;
    }

    // Get conflicts for detailed info
    const conflicts = await SyncQueue.find({
      userId,
      status: 'conflict',
    })
      .sort({ createdAt: -1 })
      .limit(10)
      .lean();

    // Phase 3.5.2: Get queue statistics
    const stats = await SyncQueue.aggregate([
      { $match: { userId: userIdObj } },
      {
        $group: {
          _id: '$dataType',
          count: { $sum: 1 },
          avgPriority: { $avg: '$priority' },
        },
      },
    ]);

    return {
      ...statusCounts,
      total: Object.values(statusCounts).reduce((sum, count) => sum + count, 0),
      conflicts: conflicts.map((item) => ({
        id: item._id,
        dataType: item.dataType,
        error: item.error,
        createdAt: item.createdAt,
        resolution: item.conflictData?.resolution,
      })),
      stats: stats.reduce((acc, stat) => {
        acc[stat._id] = {
          count: stat.count,
          avgPriority: stat.avgPriority.toFixed(2),
        };
        return acc;
      }, {}),
    };
  } catch (error) {
    logger.error('Get queue status error:', error);
    throw error;
  }
};

/**
 * Clear old synced items (cleanup)
 * Phase 3.5.2: Enhanced cleanup with configurable retention
 * @param {number} daysOld - Delete items older than X days
 */
export const clearOldSyncedItems = async (daysOld = 7) => {
  try {
    const cutoffDate = new Date();
    cutoffDate.setDate(cutoffDate.getDate() - daysOld);

    // Phase 3.5.2: Also clean up old failed items (after longer retention)
    const failedCutoffDate = new Date();
    failedCutoffDate.setDate(failedCutoffDate.getDate() - (daysOld * 2)); // Keep failed items longer

    const syncedResult = await SyncQueue.deleteMany({
      status: 'synced',
      syncedAt: { $lt: cutoffDate },
    });

    const failedResult = await SyncQueue.deleteMany({
      status: 'failed',
      updatedAt: { $lt: failedCutoffDate },
    });

    logger.info(`Cleared ${syncedResult.deletedCount} old synced items and ${failedResult.deletedCount} old failed items`);
    return {
      synced: syncedResult.deletedCount,
      failed: failedResult.deletedCount,
      total: syncedResult.deletedCount + failedResult.deletedCount,
    };
  } catch (error) {
    logger.error('Clear old synced items error:', error);
    throw error;
  }
};

/**
 * Phase 3.5.2: Get queue statistics for monitoring
 */
export const getQueueStatistics = async () => {
  try {
    const stats = await SyncQueue.aggregate([
      {
        $group: {
          _id: '$status',
          count: { $sum: 1 },
          avgPriority: { $avg: '$priority' },
          avgRetryCount: { $avg: '$retryCount' },
        },
      },
    ]);

    const byType = await SyncQueue.aggregate([
      {
        $group: {
          _id: { status: '$status', dataType: '$dataType' },
          count: { $sum: 1 },
        },
      },
    ]);

    const oldestPending = await SyncQueue.findOne({ status: 'pending' })
      .sort({ createdAt: 1 })
      .lean();

    return {
      byStatus: stats.reduce((acc, stat) => {
        acc[stat._id] = {
          count: stat.count,
          avgPriority: stat.avgPriority?.toFixed(2) || 0,
          avgRetryCount: stat.avgRetryCount?.toFixed(2) || 0,
        };
        return acc;
      }, {}),
      byType,
      oldestPending: oldestPending ? {
        id: oldestPending._id,
        createdAt: oldestPending.createdAt,
        age: Math.floor((Date.now() - new Date(oldestPending.createdAt).getTime()) / (1000 * 60 * 60)), // hours
      } : null,
    };
  } catch (error) {
    logger.error('Get queue statistics error:', error);
    throw error;
  }
};

File: mobile/lib/core/services/sync_queue_service.dart
/// Phase 3.4.0: Sync Queue Service
/// Manages sync queue with priority, retry, and persistence

import 'dart:async';
import 'package:hive_flutter/hive_flutter.dart';
import 'storage_service.dart';
import 'api_service.dart';
import '../constants/api_endpoints.dart';
import '../../features/sync/models/sync_queue_model.dart';
import 'package:connectivity_plus/connectivity_plus.dart';
import 'package:flutter/foundation.dart';
import 'package:device_info_plus/device_info_plus.dart';

class SyncQueueService {
  final StorageService _storageService;
  final ApiService _apiService;
  final Connectivity _connectivity = Connectivity();

  static const String _syncQueueBoxName = 'syncQueueBox';
  Box<dynamic>? _syncQueueBox;

  Timer? _backgroundSyncTimer;
  bool _isProcessing = false;
  StreamController<SyncQueueStatus>? _statusController;

  SyncQueueService({
    StorageService? storageService,
    ApiService? apiService,
  })  : _storageService = storageService ?? StorageService(),
        _apiService = apiService ?? ApiService();

  /// Initialize sync queue storage
  Future<void> initialize() async {
    _syncQueueBox = await _storageService.openBox(_syncQueueBoxName);
  }

  /// Get sync queue status stream
  Stream<SyncQueueStatus> get statusStream {
    _statusController ??= StreamController<SyncQueueStatus>.broadcast();
    return _statusController!.stream;
  }

  /// Check if device is online
  Future<bool> _isOnline() async {
    try {
      final connectivityResult = await _connectivity.checkConnectivity();
      return connectivityResult != ConnectivityResult.none;
    } catch (e) {
      return false;
    }
  }

  /// Get device metadata
  Future<SyncMetadata> _getMetadata() async {
    try {
      final deviceInfo = DeviceInfoPlugin();
      String? deviceId;
      
      if (defaultTargetPlatform == TargetPlatform.android) {
        final androidInfo = await deviceInfo.androidInfo;
        deviceId = androidInfo.id;
      } else if (defaultTargetPlatform == TargetPlatform.iOS) {
        final iosInfo = await deviceInfo.iosInfo;
        deviceId = iosInfo.identifierForVendor;
      }

      // Get app version from constants (package_info_plus not needed)
      return SyncMetadata(
        deviceId: deviceId,
        appVersion: '1.0.0', // Use from AppConstants if needed
        syncVersion: '1.0',
        createdAt: DateTime.now(),
      );
    } catch (e) {
      debugPrint('Error getting metadata: $e');
      return SyncMetadata(
        appVersion: '1.0.0',
        syncVersion: '1.0',
        createdAt: DateTime.now(),
      );
    }
  }

  /// Add item to sync queue
  Future<String> addToQueue({
    required String dataType,
    required Map<String, dynamic> payload,
    int priority = 5,
    int maxRetries = 3,
    DateTime? createdAt,
  }) async {
    try {
      await initialize();
      final metadata = await _getMetadata();

      final item = SyncQueueItem(
        id: '${dataType}_${DateTime.now().millisecondsSinceEpoch}_${payload.hashCode}',
        dataType: dataType,
        priority: priority,
        status: 'pending',
        payload: payload,
        maxRetries: maxRetries,
        metadata: metadata.copyWith(createdAt: createdAt ?? DateTime.now()),
        createdAt: createdAt ?? DateTime.now(),
        updatedAt: DateTime.now(),
      );

      await _syncQueueBox!.put(item.id, item.toJson());

      debugPrint('✅ Added ${dataType} to sync queue (priority: $priority, id: ${item.id})');

      // Notify listeners
      _notifyStatusUpdate();

      return item.id;
    } catch (e) {
      debugPrint('❌ Error adding to sync queue: $e');
      rethrow;
    }
  }

  /// Get pending items from queue (ordered by priority)
  Future<List<SyncQueueItem>> getPendingItems({int limit = 10}) async {
    try {
      await initialize();
      final allItems = <SyncQueueItem>[];

      for (var key in _syncQueueBox!.keys) {
        final data = _syncQueueBox!.get(key);
        if (data != null && data is Map) {
          try {
            final item = SyncQueueItem.fromJson(Map<String, dynamic>.from(data));
            if (item.status == 'pending') {
              allItems.add(item);
            }
          } catch (e) {
            debugPrint('Error parsing queue item: $e');
          }
        }
      }

      // Sort by priority (lower = higher priority), then by creation time
      allItems.sort((a, b) {
        final priorityCompare = a.priority.compareTo(b.priority);
        if (priorityCompare != 0) return priorityCompare;
        return a.createdAt.compareTo(b.createdAt);
      });

      return allItems.take(limit).toList();
    } catch (e) {
      debugPrint('Error getting pending items: $e');
      return [];
    }
  }

  /// Get conflicts
  Future<List<SyncQueueItem>> getConflicts() async {
    try {
      await initialize();
      final conflicts = <SyncQueueItem>[];

      for (var key in _syncQueueBox!.keys) {
        final data = _syncQueueBox!.get(key);
        if (data != null && data is Map) {
          try {
            final item = SyncQueueItem.fromJson(Map<String, dynamic>.from(data));
            if (item.status == 'conflict') {
              conflicts.add(item);
            }
          } catch (e) {
            debugPrint('Error parsing conflict item: $e');
          }
        }
      }

      conflicts.sort((a, b) => (a.createdAt).compareTo(b.createdAt));
      return conflicts;
    } catch (e) {
      debugPrint('Error getting conflicts: $e');
      return [];
    }
  }

  /// Process sync queue
  Future<SyncResult> processQueue({int batchSize = 10, bool force = false}) async {
    if (_isProcessing && !force) {
      return SyncResult(
        success: false,
        message: 'Queue processing already in progress',
      );
    }

    _isProcessing = true;

    try {
      if (!await _isOnline()) {
        return SyncResult(
          success: false,
          message: 'No internet connection',
        );
      }

      final pendingItems = await getPendingItems(limit: batchSize);
      if (pendingItems.isEmpty) {
        _isProcessing = false;
        return SyncResult(
          success: true,
          message: 'No pending items to sync',
          processed: 0,
          synced: 0,
          failed: 0,
        );
      }

      // Group items by data type
      final groupedByType = <String, List<SyncQueueItem>>{};
      for (final item in pendingItems) {
        groupedByType.putIfAbsent(item.dataType, () => []).add(item);
      }

      int processed = 0;
      int synced = 0;
      int failed = 0;
      final errors = <String>[];

      // Process each type
      for (final entry in groupedByType.entries) {
        final dataType = entry.key;
        final items = entry.value;

        // Update status to processing
        for (final item in items) {
          await _updateItemStatus(item.id, 'processing', retryCount: item.retryCount + 1);
        }

        try {
          // Prepare sync payload
          final syncPayload = <String, dynamic>{
            'useQueue': false, // Direct sync
          };

          if (dataType == 'quiz') {
            syncPayload['quizzes'] = items.map((item) => item.payload).toList();
          } else if (dataType == 'drill') {
            syncPayload['drillLogs'] = items.map((item) => item.payload).toList();
          } else if (dataType == 'game') {
            syncPayload['games'] = items.map((item) => item.payload).toList();
          }

          // Sync to backend
          final response = await _apiService.post(
            ApiEndpoints.sync,
            data: syncPayload,
          );

          if (response.statusCode == 200 || response.statusCode == 201) {
            final result = response.data['data'] ?? response.data;

            // Update items based on result
            for (final item in items) {
              processed++;
              bool itemSynced = false;

              if (dataType == 'quiz') {
                final syncedCount = (result['quizzes']?['synced'] as int?) ?? 0;
                if (syncedCount > 0) {
                  itemSynced = true;
                }
              } else if (dataType == 'drill') {
                final syncedCount = (result['drillLogs']?['synced'] as int?) ?? 0;
                if (syncedCount > 0) {
                  itemSynced = true;
                }
              } else if (dataType == 'game') {
                final syncedCount = (result['games']?['synced'] as int?) ?? 0;
                if (syncedCount > 0) {
                  itemSynced = true;
                }
              }

              if (itemSynced) {
                await _updateItemStatus(item.id, 'synced', syncedAt: DateTime.now());
                synced++;
              } else {
                // Check if max retries reached
                if (item.retryCount >= item.maxRetries) {
                  await _updateItemStatus(item.id, 'failed', error: 'Max retries exceeded');
                  failed++;
                } else {
                  await _updateItemStatus(item.id, 'pending');
                }
              }
            }
          } else {
            throw Exception(response.data['message'] ?? 'Sync failed');
          }
        } catch (e) {
          debugPrint('Error syncing ${dataType}: $e');
          // Update items based on retry count
          for (final item in items) {
            processed++;
            if (item.retryCount >= item.maxRetries) {
              await _updateItemStatus(item.id, 'failed', error: e.toString());
              failed++;
            } else {
              await _updateItemStatus(item.id, 'pending', error: e.toString());
            }
          }
          errors.add('${dataType}: ${e.toString()}');
        }
      }

      _notifyStatusUpdate();

      return SyncResult(
        success: synced > 0,
        message: 'Processed $processed items: $synced synced, $failed failed',
        processed: processed,
        synced: synced,
        failed: failed,
        errors: errors,
      );
    } catch (e) {
      debugPrint('Error processing queue: $e');
      return SyncResult(
        success: false,
        message: 'Queue processing error: $e',
      );
    } finally {
      _isProcessing = false;
    }
  }

  /// Update item status
  Future<void> _updateItemStatus(
    String itemId,
    String status, {
    int? retryCount,
    DateTime? syncedAt,
    String? error,
  }) async {
    try {
      await initialize();
      final data = _syncQueueBox!.get(itemId);
      if (data != null && data is Map) {
        final item = SyncQueueItem.fromJson(Map<String, dynamic>.from(data));
        final updated = item.copyWith(
          status: status,
          retryCount: retryCount ?? item.retryCount,
          syncedAt: syncedAt ?? item.syncedAt,
          error: error ?? item.error,
          lastAttemptAt: DateTime.now(),
          updatedAt: DateTime.now(),
        );
        await _syncQueueBox!.put(itemId, updated.toJson());
      }
    } catch (e) {
      debugPrint('Error updating item status: $e');
    }
  }

  /// Resolve conflict
  Future<bool> resolveConflict({
    required String itemId,
    required String resolution, // 'server-wins', 'client-wins', 'merge'
    Map<String, dynamic>? resolvedData,
  }) async {
    try {
      if (!await _isOnline()) {
        throw Exception('No internet connection');
      }

      final response = await _apiService.post(
        ApiEndpoints.resolveConflict(itemId),
        data: {
          'resolution': resolution,
          if (resolvedData != null) 'resolvedData': resolvedData,
        },
      );

      if (response.statusCode == 200) {
        // Remove from local queue if resolved
        await _removeItem(itemId);
        _notifyStatusUpdate();
        return true;
      }

      return false;
    } catch (e) {
      debugPrint('Error resolving conflict: $e');
      return false;
    }
  }

  /// Remove item from queue
  Future<void> _removeItem(String itemId) async {
    try {
      await initialize();
      await _syncQueueBox!.delete(itemId);
    } catch (e) {
      debugPrint('Error removing item: $e');
    }
  }

  /// Get queue status
  Future<SyncQueueStatus> getQueueStatus() async {
    try {
      await initialize();
      final pending = await getPendingItems(limit: 1000);
      final conflicts = await getConflicts();

      int processing = 0;
      int synced = 0;
      int failed = 0;

      for (var key in _syncQueueBox!.keys) {
        final data = _syncQueueBox!.get(key);
        if (data != null && data is Map) {
          try {
            final item = SyncQueueItem.fromJson(Map<String, dynamic>.from(data));
            if (item.status == 'processing') processing++;
            if (item.status == 'synced') synced++;
            if (item.status == 'failed') failed++;
          } catch (e) {
            // Skip invalid items
          }
        }
      }

      return SyncQueueStatus(
        pending: pending.length,
        processing: processing,
        synced: synced,
        failed: failed,
        conflict: conflicts.length,
        total: pending.length + processing + synced + failed + conflicts.length,
        conflicts: conflicts.map((c) => ConflictItem(
          id: c.id,
          dataType: c.dataType,
          error: c.error,
          createdAt: c.createdAt,
        )).toList(),
      );
    } catch (e) {
      debugPrint('Error getting queue status: $e');
      return SyncQueueStatus();
    }
  }

  /// Start background sync (process queue periodically)
  void startBackgroundSync({Duration interval = const Duration(minutes: 5)}) {
    _backgroundSyncTimer?.cancel();
    _backgroundSyncTimer = Timer.periodic(interval, (_) async {
      if (await _isOnline() && !_isProcessing) {
        await processQueue();
      }
    });
    debugPrint('✅ Background sync queue processing started');
  }

  /// Stop background sync
  void stopBackgroundSync() {
    _backgroundSyncTimer?.cancel();
    _backgroundSyncTimer = null;
    debugPrint('⏹️  Background sync queue processing stopped');
  }

  /// Notify status update
  void _notifyStatusUpdate() {
    getQueueStatus().then((status) {
      _statusController?.add(status);
    });
  }

  /// Clear old synced items (cleanup)
  Future<int> clearOldSyncedItems({int daysOld = 7}) async {
    try {
      await initialize();
      final cutoffDate = DateTime.now().subtract(Duration(days: daysOld));
      int cleared = 0;

      for (var key in _syncQueueBox!.keys) {
        final data = _syncQueueBox!.get(key);
        if (data != null && data is Map) {
          try {
            final item = SyncQueueItem.fromJson(Map<String, dynamic>.from(data));
            if (item.status == 'synced' && item.syncedAt != null && item.syncedAt!.isBefore(cutoffDate)) {
              await _syncQueueBox!.delete(key);
              cleared++;
            }
          } catch (e) {
            // Skip invalid items
          }
        }
      }

      if (cleared > 0) {
        _notifyStatusUpdate();
      }

      return cleared;
    } catch (e) {
      debugPrint('Error clearing old items: $e');
      return 0;
    }
  }

  /// Dispose resources
  void dispose() {
    stopBackgroundSync();
    _statusController?.close();
    _statusController = null;
  }
}

class SyncResult {
  final bool success;
  final String message;
  final int processed;
  final int synced;
  final int failed;
  final List<String> errors;

  SyncResult({
    required this.success,
    required this.message,
    this.processed = 0,
    this.synced = 0,
    this.failed = 0,
    this.errors = const [],
  });
}

